x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -736,7 +802,218 @@ public void runAsDenied(String requestId, Authentication authentication, RestReq
+            logEntry.with(PUT_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(PutRoleMappingRequest putRoleMappingRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "put_role_mapping");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("role_mapping")
+                    .field("name", putRoleMappingRequest.getName())
+                    .field("role_names", putRoleMappingRequest.getRoles())
+                    .field("role_templates", putRoleMappingRequest.getRoleTemplates())
+                    .field("rules", putRoleMappingRequest.getRules())
+                    .field("enabled", putRoleMappingRequest.isEnabled())
+                    .field("metadata", putRoleMappingRequest.getMetadata())
+                    .endObject() // role_mapping
+                    .endObject();
+            logEntry.with(PUT_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(SetEnabledRequest setEnabledRequest) throws IOException {
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("user")
+                    .field("name", setEnabledRequest.username())
+                    .field("realm", inferRealmNameFromUsername.apply(setEnabledRequest.username()))
+                    .endObject() // user
+                    .endObject();
+            if (setEnabledRequest.enabled()) {
+                logEntry.with(EVENT_ACTION_FIELD_NAME, "enable_user");
+                logEntry.with(ENABLE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            } else {
+                logEntry.with(EVENT_ACTION_FIELD_NAME, "disable_user");
+                logEntry.with(DISABLE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            }
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(PutPrivilegesRequest putPrivilegesRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "put_privileges");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .field("privileges", putPrivilegesRequest.getPrivileges())
+                    .endObject();
+            logEntry.with(PUT_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(CreateApiKeyRequest createApiKeyRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "create_apikey");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            TimeValue expiration = createApiKeyRequest.getExpiration();
+            builder.startObject()
```

Comment:
api key `id` not currently audited we need to settle on the best approach, between carrying the `id` in the request object and intercepting the `id` from the response (see https://github.com/elastic/elasticsearch/pull/63708).